### PR TITLE
Clean up Cell Clearing

### LINKED ISSUE #646: Compiling an empty cell doesn't clear the earlier outputs
Even when I compile an empty cell, the earlier outputs doesn't get cleared.

Though a trivial thing but looks odd in the notebook.

### LINKED ISSUE #650: clear_output not supported in notebook
`clear_output` is used to do in-place updates of information, here's code for a simple example:

```python
from time import sleep
from IPython.display import clear_output

n, go = 0, True
while go:
    try:
        clear_output(wait=True)
        print("n=", n)
        n += 1
        sleep(0.3)
    except KeyboardInterrupt:
        go = False
        print("bye!")
```
The current nb provides this behavior:

nb-clear-output

whereas in JLab we aren't getting the in-place clearing of old output:

jlab-clear-output

### PR DESCRIPTION
**Pull request by blink1073**
Fixes #650.  Fixes #646.

screen shot 2016-08-13 at 3 41 56 am

### CODE REVIEW

File: src/notebook/output-area/model.ts [hunk 1]
```diff
@@ -177,6 +177,9 @@ class OutputAreaModel implements IDisposable {
           model.output_type = msgType;
           this.add(model);
           break;
+        case 'clear_output':
+          this.clear(true);
```
Comment (jasongrout) [inline src/notebook/output-area/model.ts]: The argument should be the value given in the clear_output message: http://jupyter-client.readthedocs.io/en/latest/messaging.html#clear-output

Comment (blink1073) [inline src/notebook/output-area/model.ts]: Nice catch!

File: src/notebook/cells/widget.ts [hunk 2]
```diff
@@ -430,6 +430,7 @@ class CodeCellWidget extends BaseCellWidget {
   execute(kernel: IKernel): Promise<KernelMessage.IExecuteReplyMsg> {
     let model = this.model as ICodeCellModel;
     let code = model.source;
+    model.outputs.clear();
```
Comment (jasongrout) [inline src/notebook/cells/widget.ts]: Isn't this already being run in the output area execution method, at https://github.com/jupyter/jupyterlab/pull/656/files#diff-1eb8c480985ee63765746a78fa427fedR165?

Comment (blink1073) [inline src/notebook/cells/widget.ts]: Good point, moved it within the `if` block.

**Comment by blink1073** [comment]
I confirmed that the `wait` parameter works now.

File: src/notebook/cells/widget.ts [hunk 3]
```diff
@@ -430,9 +430,9 @@ class CodeCellWidget extends BaseCellWidget {
   execute(kernel: IKernel): Promise<KernelMessage.IExecuteReplyMsg> {
     let model = this.model as ICodeCellModel;
     let code = model.source;
-    model.outputs.clear();
     if (!code.trim()) {
       model.executionCount = null;
+      model.outputs.clear();
```
Comment (jasongrout) [inline src/notebook/cells/widget.ts]: Ah, much clearer. Thanks!

**Pull request merged**